### ISSUE TITLE
MariaDB 10.2.6 error in handshake

### ISSUE THREAD

**Issue by ea7ababe** [mysql-otp/mysql-otp#72]
Hello there,

today I have updated MariaDB from 10.1.24 to 10.2.6 and now `mysql:start_link([{name, {local,db}},{database,D},{user,U},{password,P}])` raises an exception:

```
** exception exit: {badmatch,2}
     in function  mysql_protocol:handshake/6 (/usr/home/u/project/_build/default/lib/mysql/src/mysql_protocol.erl, line 60)
     in call from mysql:init/1 (/usr/home/u/project/_build/default/lib/mysql/src/mysql.erl, line 497)
     in call from gen_server:init_it/6 (gen_server.erl, line 328)
     in call from proc_lib:init_p_do_apply/3 (proc_lib.erl, line 247)
```

When I comment out that line 60 in mysql_protocol.erl everything seems to work fine...

**Comment by ionuthristodorescu** [comment]
Same here, on macOS Sierra 10.12.6. Has anybody been able to look into this issue ? It fails out this assert at line 60 : "OkStatus = Handshake#handshake.status" because OkStatus is 16386 (0x4002 in hex) and handshake.status is 2.

**Comment by ionuthristodorescu** [comment]
So, looks like for Status fields (https://dev.mysql.com/doc/dev/mysql-server/8.0.0/mysql__com_8h.html), 0x4002 is actually SERVER_STATUS_AUTOCOMMIT | SERVER_SESSION_STATE_CHANGED. The problem is, SERVER_SESSION_STATE_CHANGED should be set in the Status flags ONLY if CLIENT_SESSION_TRACK is set - but it says in the comments in mysql_protocol.erl that "We have CLIENT_PROTOCOL_41 but not CLIENT_SESSION_TRACK enabled". I'll continue debugging and update here.

**Comment by ionuthristodorescu** [comment]
https://mariadb.com/kb/en/mariadb/1-connecting-connecting/  : 

if (server_capabilities & CLIENT_MYSQL)
  string<4> filler
else
  int<4> server capabilities 3rd part . MariaDB specific flags /* MariaDB 10.2 or later */

**Comment by sfilargi** [comment]
Is there a workaround?

**Comment by sfilargi** [comment]
Should we be looking at the client flags, or the server flags?
I.e.

```
#define SERVER_STATUS_ANSI_QUOTES       32768U
```

**Comment by sfilargi** [comment]
> The problem is, SERVER_SESSION_STATE_CHANGED should be set in the Status flags ONLY if CLIENT_SESSION_TRACK is set

I think this is misinterpretation. MariaDB will reply with whatever the server_status is, no matter if CLIENT_SESSION_TRACK is enabled or not. The CLIENT_SESSION_TRACK only affects if the extra fields get added in the OK packet(https://mariadb.com/kb/en/library/ok_packet/).


We can see this in the code:
First it sets that status as is:
```
store.q_append2b(server_status);
```

And then it checks for CLIENT_SESSION_TRACK, in order to include the extra optional info:
```
state_changed=
    (thd->client_capabilities & CLIENT_SESSION_TRACK) &&
    (server_status & SERVER_SESSION_STATE_CHANGED);

  if (state_changed || (message && message[0]))
  {
    DBUG_ASSERT(safe_strlen(message) <= MYSQL_ERRMSG_SIZE);
    store.q_net_store_data((uchar*) safe_str(message), safe_strlen(message));
}
```

https://github.com/MariaDB/server/blob/8f643e2063c9890a353149f39ef85b2cf3151fd0/sql/protocol.cc#L256

**Comment by sfilargi** [comment]
**Workaround**

If you don't specify a database at the time of the connection, then it should work.

So do something like this instead:
```
{ok, Pid} = mysql:start_link([{host, "localhost"}, {user, "foo"},
                              {password, "hello"}, {ssl, [{cacertfile, "/path/to/ca.pem"}]}]),
 ok = mysql:query(Pid, "USE test"),
```

**Comment by zuiderkwast** [comment]
Is there a way to fix this without removing the check that the flags are what we expect them to be? In a way that is good for both old and new mysql and mariadb?

**Comment by emqplus** [comment]
@ea7ababe @zuiderkwast @ionuthristodorescu @sfilargi I submit PR #75 to fix the issue, please help review it.

**Comment by zuiderkwast** [comment]
Closed in PR #75
